reframe/core/parameters.py

```diff
@@ -0,0 +1,200 @@
+                    raise NameConflictError(f'parameter space conflict '
+                                            f'(on {key}) due to '
+                                            f'multiple inheritance.'
+                                           ) from None
+
+            param_space[key] = base_params.get(
+                key, ()) + param_space.get(key, ())
+
+    return param_space
+
+
+def _extend_parameter_space(local_param_space, param_space):
+    '''Extend a given parameter space with a local parameter space.
+
+    Each parameter is dealt with independently, given that each parameter
+    has its own inheritance behaviour defined in the local parameter space
+    (see the
+    :class:`reframe.core.parameters_TestParameter` class).
+
+    :param local_param_space: a local parameter space from a regression test.
+        This must be an instance of the class
+        :class:`reframe.core.parameters.LocalParamSpace`.
+    :param param_space: an existing parameter space. This **must** have been
+        generated with
+        :meth:`reframe.core.parameters._merge_parameter_spaces`.
+
+    '''
+    # Ensure that the local parameter space is an instance of LocalParamSpace
+    if not isinstance(local_param_space, LocalParamSpace):
+        raise ReframeSyntaxError(f'the local_param_space must be an '
+                                 f'instance of the LocalParamSpace class')
+
+    # Loop over the local parameter space.
+    for name, p in local_param_space.params.items():
+        param_space[name] = p.filter_params(
+            param_space.get(name, ())) + p.values
+
+
+def build_parameter_space(cls):
+    ''' Builder of the full parameter space of a regression test.
+
+    Handles the full parameter space build, inheriting the parameter spaces
+    form the base classes, and extending these with the local parameter space
+    of the class (stored in cls._rfm_local_param_space). This method is called
+    during the class object initialization (i.e. the __init__ method of the
+    regression test metaclass). This method has three main steps, which are
+    (in order of execution) the inheritance of the parameter spaces from the
+    base classes, the extension of the inherited parameter space with the
+    local parameter space, and lastly, a check to ensure that none of the
+    parameter names clashes with any of the class attributes existing in the
+    regression test class.
+
+    :param cls: the class where the full parameter space is to be built.
+
+    :returns: dictionary containing the full parameter space. The keys are the
+        parameter names and the values are tuples containing all the values
+        for each of the parameters.
+    '''
+    # Inherit the parameter space from the base classes
+    param_space = _merge_parameter_spaces(cls.__bases__)
+
+    # Extend the parameter space with the local parameter space
+    _extend_parameter_space(cls._rfm_local_param_space, param_space)
```

Comment:
Using `cls.mro()` wouldn't work here. `cls.mro()` returns the full hierarchy of parent classes, and that would trigger the `parameter space conflict (on {key}) due to multiple inheritance` error. There is no need to pass the full inheritance tree in here because the direct parent has already it's own `cls._rfm_params` built in there. So just passing the direct parents with `cls.__bases__` is what we want here.